### PR TITLE
Make result iteration recoverable from errors

### LINKED ISSUE #598: TrinoResult generator cannot recover from errors during lazy segment iteration
Python generators cannot be resumed after an unhandled exception, the frame is finalized and all subsequent next() calls raise `StopIteration`.

After #597 spooled segment downloads happen lazily inside the generator during iteration. If a transient error (e.g. S3 timeout) occurs mid-iteration, the caller sees the exception, but the generator is permanently dead.
Subsequent fetchone() calls also return None as if the query completed normally and silently drops remaining rows.

Before #597 this was theoretical issue since fetch() materialized all segments eagerly. I/O errors happened outside the generator.

From looking at other projects it seems the pattern is to use a class with `__next__` and `__iter__` instead of a generator and having a field on the class to hold state of the iterator so that we can signal to the caller what happened.

Here's a minimal repro:

```python
from trino.client import TrinoResult

class FailingIterator:
    def __init__(self):
        self._count = 0
    def __iter__(self):
        return self
    def __next__(self):
        if self._count >= 3:
            raise IOError("S3 segment download failed")
        self._count += 1
        return [self._count]

class FakeQuery:
    finished = True
    def fetch(self):
        return []

result = TrinoResult(FakeQuery(), FailingIterator())
it = iter(result)
```

```
next(it)  # [1]
next(it)  # [2]
next(it)  # [3]
next(it)  # raises IOError, caller sees this and retries
next(it)  # raises StopIteration, caller gets None from fetchone() and thinks query is done
```

### PR DESCRIPTION
**Pull request by hashhar**
## Description

TrinoResult implemented iteration as a generator. A generator that raises an unhandled exception is finalized so every subsequent next() raises StopIteration which dbapi's fetchone() maps to None. Since spooled segments are downloaded lazily during iteration, a transient error (e.g. an S3 timeout) mid-iteration permanently killed the iterator and silently dropped the remaining rows as if the query had completed normally.

Rewrite TrinoResult as a class-based iterator that keeps its state on the instance. An error propagates to the caller while the iterator stays usable, so a persistent error keeps surfacing instead of turning into StopIteration and a retried next() resumes where the failure happened.

Also make SegmentIterator retry-safe. Acknowledge and advance past a segment only after it was decoded successfully so retrying after a failed segment download re-decodes the same segment instead of acknowledging and skipping it.

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
* Fix some things. ({issue}`issuenumber`)
```

Fixes #598.

### CODE REVIEW

File: tests/unit/test_client.py [hunk 1]
```diff
@@ -1448,3 +1448,92 @@ def test_decoder_factory_raises_with_message_on_missing_lz4():
...
+            self._served = 0
+            self._failed = False
+
+        def __iter__(self):
+            return self
+
+        def __next__(self):
+            if self._served == 2 and not self._failed:
+                self._failed = True
+                raise IOError("segment download failed")
```
Comment (azawlocki-sbdt) [inline tests/unit/test_client.py]: Maybe also add a test case for ensuring that `TrinoResult` can recover from two or three consecutive errors in the underlying iterator, not just a single one.

File: tests/unit/test_client_spooling.py [hunk 2]
```diff
@@ -200,3 +200,52 @@ def test_fetch_passes_request_and_interval_to_segment_iterator(heartbeat_interva
...
+            self._failing_segment = None
+            raise IOError("segment download failed")
+        return self._rows_by_segment[segment]
+
+
+def test_segment_iterator_retries_failed_segment_without_skipping_it():
+    seg1, seg2, seg3 = (_FakeSpooledSegment(name) for name in ("s1", "s2", "s3"))
+    segments = [DecodableSegment("json", None, seg) for seg in (seg1, seg2, seg3)]
+    iterator = SegmentIterator(segments, mapper=None)
+    iterator._decoder = _FlakyDecoder({seg1: [[1]], seg2: [[2]], seg3: [[3]]}, failing_segment=seg2)
```
Comment (azawlocki-sbdt) [inline tests/unit/test_client_spooling.py]: I think the first/last segments failing are corner cases worth covering. Maybe parametrize this test with `failing_segment_index=0,1,2`?

File: trino/client.py [hunk 3]
```diff
@@ -844,15 +853,28 @@ def rownumber(self) -> int:
...
 
     def __iter__(self):
-        # A query only transitions to a FINISHED state when the results are fully consumed:
-        # The reception of the data is acknowledged by calling the next_uri before exposing the data through dbapi.
-        while not self._query.finished or self._rows is not None:
-            next_rows = self._query.fetch() if not self._query.finished else None
-            for row in self._rows:
-                self._rownumber += 1
-                yield row
+        return self
```
Comment (azawlocki-sbdt) [inline trino/client.py:856]: NB: Side effect of the changes is that now we have `it_1 = iter(trino_result); it_2 = iter(trino_result); it_1 is it_2`. Previously we had `it_1 is not it_2`, but `it_1` and `it_2` shared the underlying result object anyway and would interfere.

**Review by azawlocki-sbdt — APPROVED** [review]
LGTM, code comments make it a good read. My only suggestion would be to cover more scenarios in tests, but it's non-blocking.
